WcaOnRails/app/webpacker/javascript/hooks/useSaveData.js

```diff
@@ -0,0 +1,35 @@
+import { useState, useCallback } from 'react';
+import { fetchJsonOrError } from '../requests/fetchWithAuthenticityToken';
+
+const throwError = (err) => { throw err; };
+
+// This is a hook that can be used to save some data to the website (as json)
+// It assumes that 'url' is a valid, PATCH-able, url.
+// Example of usage:
+// const { save, saving } = useSaveData();
+// // or if you want to override some options:
+// const { save, saving } = useSaveData({ method: 'DELETE' });
+// // and then:
+// save(modelUrl(), { /* model attrs */ }, () => console.log("success"));
+const useSaveData = (options = {}) => {
```

Comment:
Why are we only passing `options` here, and no `url`?
In `useLoadedData`, the URL gets passed directly into the hook "constructor" whereas here it has to be added upon each individual invocation of the `save` method.

Reply:
Because `useLoadedData` is used to sync one single ressource, and the request is made upon creating the hook, whereas `useSaveData` returns a method that can be used multiple times to save multiple ressources.
Setting the url here would prevent us from reusing the hook [here](https://github.com/thewca/worldcubeassociation.org/pull/6360/files#diff-16c23ab34d8ba9d3740c1cfe4faa6b4a923434df095974fe431c048d162d05eaR118) for instance.

But the point about options could actually be made the other way around: we should be able to override the options when calling `save`!
This would actually let us use `save` for the [`deleteAction` here](https://github.com/thewca/worldcubeassociation.org/pull/6360/files#diff-16c23ab34d8ba9d3740c1cfe4faa6b4a923434df095974fe431c048d162d05eaR128).

Reply:
I see your point, but I am not yet convinced by your argument. If the goal here is reusing the hook, then `useLoadedData` should have the ability to invoke multiple URLs with a single hook, too. At least as long as these two hooks are semantically close/related.

I just find it unintuitive to have two hooks which are so similar in nature (one _loads_ data, one _saves_ data) but so fundamentally different in usage.
Say you have a page that uses (fictional) `User` data and `Result` data to do some (fictional) WRT work to reassign results to different users. Reading the user and the results affiliated with that user would require individual hooks each, but the saving mechanism would just require one hook that is re-used across models. To me, this would be mildly confusing, in a sense that I could get used to it but I would probably scratch my nose every time I create those hooks.

On that note, why is it so desirable to reuse the save hook but not the read hook? If we do end up going down that path, then why even provide the ability to pass the HTTP verb (via `options`) in the hook "constructor" in the first place? Imagine a page with a "create or update" functionality, that would require `PUT` and `PATCH` hooks, respectively. You'd have to create two hooks and choose which one to use anyways. I see you're doing [this](https://github.com/thewca/worldcubeassociation.org/pull/6360/files#diff-16c23ab34d8ba9d3740c1cfe4faa6b4a923434df095974fe431c048d162d05eaR188) in your other PR, but doesn't that contradict reusability because every individual wrapper then computes its own hook anyways?

Sorry for the wall of text, I just want to understand why it's desirable to reuse this hook but not `useLoadedData`.

Reply:
Kindly ping @viroulep :blush:

Reply:
> If the goal here is reusing the hook, then useLoadedData should have the ability to invoke multiple URLs with a single hook, too. At least as long as these two hooks are semantically close/related.

No it can't: `useLoadedData` provides two features:
  - an async load that starts as soon as the hook is created
  - a `sync` feature that can be manually triggered by the user at any point in time, that reloads the data and update the provided `data`.

Because of this it simply makes no sense to change the entity tied to the hook "on the fly".

Maybe the name is not chosen appropriately, but the intent of `useSaveData` is simply to provide an easy way to update an object on the server: you give it a url, some data, some callback, and it handles the rest for you.
Once the request is performed it doesn't have any other features related to the object saved to offer.

> If we do end up going down that path, then why even provide the ability to pass the HTTP verb (via options) in the hook "constructor" in the first place?

Yes that's a good point and actually I've just pushed a change where I moved "down" the `options` parameter!

[Here](https://github.com/thewca/worldcubeassociation.org/pull/6360/files#diff-16c23ab34d8ba9d3740c1cfe4faa6b4a923434df095974fe431c048d162d05eaR132) it actually allows to use only a single hook for creating, patching, or deleting a result.
In the diff you mentioned I was able to change the request method in the constructor for creating/patching, but it was not possible to reuse it beyond that and I needed to use another hook/a direct call to fetch in order to delete the result.

If it helps we can rename this `useFetchQuery` or something, but I really don't see the point of tying the hook to an entity when it doesn't actually need to.

If we want something properly tied to an object we should have a hook that exposes some features to transparently "mutate" the object, and which probably need to rely on the server API to have some good conventions about how PATCH/POST/DELETE routes are created.
To be honest this sounds like a different beast to me, and most likely the point where we should use [react-query](https://react-query.tanstack.com/reference/useMutation).

Reply:
Thanks for the thorough feedback! Your point on `useLoadedData` is convincing.

Also, I realised this morning that there is a crucial difference in terms of names between `useSaveData` and `useSav*ed*Data`. The hook for loading makes sense because you're referencing something that is _being loaded_ by the hook itself, whereas for saving you can use the other hook when you're _about to save_ something.

I acknowledge that this minute difference of the `-ed` suffix makes a significant difference in terms of English. But it still looks too similar in code. Perhaps we can resolve our creative differences by finding a name for the save hook that does not sound too similar to the load hook? My suggestions are:
- `useMutation`
- `useSaveDelegate`, `useSaveService`
- `useWriteDelegate`, `useWriteService`
As you can see, I'm a bit unsure whether what we're doing here is actually a Delegate or a Service in terms of patterns, but it certainly *is not a model state*. It is something that *helps you act on model state*, as opposed to the load hook which actually provides a model state (along with some helpers)  

Of course the ultimate goal should be to use a widely adopted third-party library. But for the sake of moving forward and not losing ourselves in even more details, I think we can merge this as soon as we agreed on a different name here.

Reply:
I agree with your explanation :)

I think `useMutation` is too close from react-query, however it definitely wouldn't provide the same things so I think we should avoid this one.
What do you think about `useSaveAction`?
It's quite close to what the hook provides (an action one can call to save data).

Reply:
I like that very much! Also cleverly avoids the design-pattern buzzword bingo :D

Reply:
Great, I'll try to update this PR and the results' one with that change rather soon (let's be optimistic and aim for tomorrow :))!